programs/perpetuals/tests/utils/utils.rs

```diff
@@ -0,0 +1,463 @@
+use std::path::Path;
+
+use anchor_lang::{prelude::*, InstructionData};
+use anchor_spl::token::spl_token;
+use bonfida_test_utils::ProgramTestContextExt;
+use perpetuals::{
+    instructions::{
+        AddCustodyParams, AddLiquidityParams, SetCustodyConfigParams, SetTestOraclePriceParams,
+    },
+    state::{
+        custody::{Custody, Fees, PricingParams},
+        perpetuals::Permissions,
+    },
+};
+use solana_program::{bpf_loader_upgradeable, program_pack::Pack, stake_history::Epoch};
+use solana_program_test::{read_file, ProgramTest, ProgramTestContext};
+use solana_sdk::{account, signature::Keypair, signer::Signer, signers::Signers};
+
+use crate::instructions;
+
+use super::{fixtures, get_program_data_pda, get_test_oracle_account};
+
+pub const ANCHOR_DISCRIMINATOR_SIZE: usize = 8;
+
+pub fn create_and_fund_account(address: &Pubkey, program_test: &mut ProgramTest) {
+    program_test.add_account(
+        *address,
+        account::Account {
+            lamports: 1_000_000_000,
+            ..account::Account::default()
+        },
+    );
+}
+
+pub fn find_associated_token_account(owner: &Pubkey, mint: &Pubkey) -> (Pubkey, u8) {
+    Pubkey::find_program_address(
+        &[
+            owner.as_ref(),
+            anchor_spl::token::ID.as_ref(),
+            mint.as_ref(),
+        ],
+        &anchor_spl::associated_token::ID,
+    )
+}
+
+pub fn copy_keypair(keypair: &Keypair) -> Keypair {
+    Keypair::from_bytes(&keypair.to_bytes()).unwrap()
+}
+
+pub async fn get_token_account(
+    program_test_ctx: &mut ProgramTestContext,
```

Comment:
why getters (this and below) take mutable arguments?

Reply:
hm, not sure why get_account() needs mut self, maybe updating a cache or something. One way to overcome this is by creating a struct that would wrap all functions that take ProgramTestContext, and ProgramTestContext will be decalared as RefCell, i.e. 
```
pub struct CoolWrapper {
   ctx: RefCell<ProgramTestContext>,
   other stuff
}

impl CoolWrapper {
  pub fn new() -> Self {
     Self { ctx: RefCell::new(ProgramTestContext::default()), }
  }
  pub fn do_something() {}
  pub fn get_something() -> Account {
     ctx.get_account(...)
  }
  ...
}
```
but that will probably require to redesign some other parts of your framework, so fine to leave as is for now